File: 100-same-tree/same-tree.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution(object):
    def isSameTree(self, p, q):
        if not p and not q:  # both None
            return True
        if not p or not q:   # one is None, the other is not
            return False
        if p.val != q.val:   # values differ
            return False
        # recursively check left and right subtrees
        return self.isSameTree(p.left, q.left) and self.isSameTree(p.right, q.right)

File: 100-same-tree/same-tree.cpp
class Solution {
private:
    bool preorder(TreeNode* p, TreeNode* q) {
        if (!p && !q) return true;  // both null => same
        if (!p || !q) return false; // one null, one not => different
        
        if (p->val != q->val) return false;
        
        bool leftsame = preorder(p->left, q->left);
        bool rightsame = preorder(p->right, q->right);
        
        return leftsame && rightsame;
    }
public:
    bool isSameTree(TreeNode* p, TreeNode* q) {
        return preorder(p, q);
    }
};
